docs: login rule docs (#21195)
gravitational/teleport: docs/pages/access-controls/login-rules/guide.mdx

---

<Admonition type="warning" title="Preview">
  Login Rules are currently in Preview mode.
</Admonition>

This guide will walk you through the process of writing, testing, and adding the
first Login Rule to your Teleport Cluster.

## Prerequisites

(!docs/pages/includes/commercial-prereqs-tabs.mdx!)

(!docs/pages/includes/tctl.mdx!)

Before you get started you’ll need a running Teleport Enterprise or Cloud
cluster on version 11.3.0 or greater.

Login Rules only have an effect when an SSO user logs in, so make sure you have
configured an OIDC, SAML, or GitHub connector before you begin.
Check the [Single Sign-On](../sso/) docs to learn how to set this up.

## Step 1/4. Draft your Login Rule resource

Each Login Rule resource must have either a `traits_map` or `traits_expression`
field.
In this guide we will use an example `traits_expression`.

The `traits_expression` is a form of script which will be evaluated by your
Teleport cluster at runtime to determine the traits for each SSO user who logs
in.
The expression can access the incoming traits for the user via the `external`
variable.
The `external` variable is a dictionary which maps trait keys to sets of values

> for that trait.

Author: moved to after the snipper

The following example Login Rule adds a `"logins"` trait set to the value of the
incoming `"username"` trait, converted to lowercase.
Copy this example rule to a file called `my_rule.yaml` to continue with the

> guide.

Author: I struggled to make this much more general, came up this for now
> The following example will give all users a new `logins` trait set to the value of their current `username` trait converted to lowercase.

As a final step, relogin to your cluster and make sure your user received the
expected traits and roles.
You can check the traits and roles with the following command:

```code
$ tctl get --format json users/<Var name="username"/> | jq '{traits: first.spec.traits, roles: first.spec.roles}'
{
  "traits": {
	"access": [
	  "staging"
    ],
    "groups": [
      "dbs",
      "devs"
    ],
    "logins": [
      "alice"
    ]
  },
  "roles": [
    "access",
    "editor",
    "auditor"
  ]
}
```

## Troubleshooting

The [`tctl sso test`](../reference/cli.mdx#tctl-sso-test) command can be used to debug SSO logins and see exactly
which traits are being sent by your SSO provider and how they are being mapped
by your Login Rules.
`tctl sso test` expects a connector spec.
To retrieve the one currently installed in your cluster, run
`tctl get saml/<connector_name> --with-secrets | tctl sso test`
(replace `saml` with `oidc` or `github` as necessary).

Run the `tctl get login_rules` command to view all Login Rules installed in your
cluster.
Run `tctl get login_rule/<rule_name>` to view a specific Login Rule.
Run `tctl create -f login_rule.yaml` to overwrite an existing Login Rule.

> Run `tctl rm login_rule/<rule_name>` to delete a Login Rule.

Author: moved these to a table in the Next Steps section

---
title: Set Up Login Rules (Preview)
description: Set up Login Rules to transform user traits
---

<Admonition type="warning" title="Preview">
  Login Rules are currently in Preview mode.
</Admonition>

This guide will walk you through the process of writing, testing, and adding the
first Login Rule to your Teleport cluster.

## Prerequisites

(!docs/pages/includes/commercial-prereqs-tabs.mdx!)

(!docs/pages/includes/tctl.mdx!)

Before you get started you’ll need a running Teleport Enterprise or Cloud
cluster on version 11.3.0 or greater.

> Login Rules only have an effect when an SSO user logs in, so make sure you have

Reviewer: ```suggestion
Login Rules only operate on SSO logins, so make sure you have
```

configured an OIDC, SAML, or GitHub connector before you begin.
Check the [Single Sign-On](../sso/) docs to learn how to set this up.

## Step 1/4. Draft your Login Rule resource

The following example will give all users a new `logins` trait set to the value
of their current `username` trait converted to lowercase.
Copy this example rule to a file called `my_rule.yaml` to continue with the
guide.

```
# my_rule.yaml
kind: login_rule
version: v1
metadata:
  # Each Login Rule must have a unique name within the cluster.
  name: my_rule

  # expires is optional and usually should not be set for deployed login
  # rules, but it can be useful to set an expiry a short time in the future
  # while testing new Login Rules to prevent potentially locking yourself out of
  # your Teleport cluster.
  # expires: "2023-01-31T00:00:00-00:00"
spec:
  # priority orders the evaluation of Login Rules if multiple are present in the
  # cluster, lower priorities are evaluated first.

>   priority: 0

Reviewer: Can priority be negative or is `0` the highest priority?

Author: it can be negative (it's an int32). I added a mention of the default value and supported range for the reference page, I think it can be left out of the guide
